tutorials/haskell/hspec-golden-test-tutorial/tutorial.md

```diff
@@ -0,0 +1,506 @@
+```
+and writting our test.
+
+```haskell
+spec :: Spec
+spec =
+    describe "renderHtml" $
+      it "Renders an Html5 file " $
+       defaultGolden "html" (htmlRendered somePage)
+```
+
+Just like the previous example, after running the test we will have a new directory `/.golden/html/` that contains our actual and golden file. But what would happen if we would have to change our HTML template ?
+
+In the practice changes may include several lines and tags but to make a simple example, let's edit the body of our HTML to "Goodbye Golden Testers" and run the tests.    
+
+```shell
+test/Html/HtmlGoldenSpec.hs:11:5:
+  1) Html.HtmlGolden.renderHtml Renders an Html5 file
+      expected: "   
+       <body>
+               Hello Golden Testers.
+       </body>"
+      but got: "
+       <body>
+               Goodbye Golden Testers.
+       </body>"
+
+
+Finished in 0.0013 seconds
+3 examples, 1 failure
+```
+
+Our HTML test failed, but not big deal, we can edit our `golden` file and things will work just fine. Pretty simple right?
+
+Actually its not, let{s remeber this is just a didactic example, in the real life golden files stores really large outputs and it would be a waste of time to update them by hand. Fortunately `hspec-golden` provides us with `hgold` CLI tool that automatically updates golden files using. Let's install it.
+
+Using stack:
+
+```shell
+$stack install hspec-golden
+```
+
+using Cabal:
+
+```shell
+$cabal install hspec-golden
+```
+Once installed we can use the --help flag to see how the CLI works.
+
+According to [hspec-golden](http://hackage.haskell.org/package/hspec-golden), when `hgold` is used without flags it updates by default the `/.golden` directory. If we store our golden files in a different directory we should use the `--update` flag and specify the name of the directory as an argument.
+
+Ok, lets update our golden files. 
+
+```shell
+$ hgold
+Replacing golden with actual...
+  Replacing file: .golden/hello/golden with: .golden/hello/actual
+  Replacing file: .golden/html/golden with: .golden/html/actual
+Finish...
+```
+
+Hgold replaces the `golden` file (Expected output) with the content of the `actual` file (SUT Output) so it is necessary to run the test first (even if they fail) to update our tests.
+
+So now we are ready to test our updated golden files.
+
+```shell 
+Hello.HelloGolden
+  sayHi
+    returns Hello Golden Testers string
+      Golden and Actual output hasn't changed
+Hello.Hello
+  sayHi
+    shows a Hello Golden Testers string
+Html.HtmlGolden
+  renderHtml
+    Renders an Html5 file 
+      Golden and Actual output hasn't changed
+
+Finished in 0.0034 seconds
+3 examples, 0 failures
+```
+
+And we can see that everything works fine.
+
+#### JSON Golden Test
+
+Until now we have managed components that returns strings only, but what if our return type is different like a ByteString or Text? For example, lets analize the `encodeCountries` function in the `Json.hs` module.
```

Comment:
@Jagl257 Let's enclose `ByteString` and `Text` with backticks.

Reply:
Ok @CristhianMotoche , done.